transactions.md

```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+  ~~~
+
+  In these cases, CockroachDB retries the transaction using the existing values. If any of the transaction's values are conditional, we recommend instead writing your transaction to use [client-side intervention](#client-side-intervention).
+
+### Client-Side Intervention
+
+In cases of contention, your application must include error handling when the transaction contains statements on separate lines, such as:
+
+~~~ sql
+> BEGIN;
+
+> UPDATE products SET inventory = 0 WHERE sku = '8675309';
+
+> INSERT INTO orders (customer, status) VALUES (1, 'new');
 
-4. If a statement returns a retryable error (identified via the `40001` error code or `restart transaction` string at the start of the error message), the [`ROLLBACK TO SAVEPOINT cockroach_restart`](rollback-transaction.html) statement restarts the transaction. Alternately, the original `SAVEPOINT cockroach_restart` statement can be reissued to restart the transaction.
+> COMMIT;
+~~~
+
+To alert you of failures due to contention, CockroachDB surfaces an error with the code `40001` that begins with the string `retry transaction`.
+
+To handle these types of errors you have two options:
+
+- *Recommended*: Use the `cockroach_restart` functions to create retryable transactions. Retryable transaction can improve performance because their priority's increased each time they are retried, making them more likely to succeed the longer they're in your system. 
+
+  For more information, see [Client-Side Transaction Retries](#client-side-transaction-retries).
+
+- Abort the transaction using `ROLLBACK`, and then reissue all of the statements in the transaction. This does *not* automatically increase the transaction's priority, so it's possible in high-contention workloads for transactions to take an incredibly long time to succeed.
+
+#### Client-Side Transaction Retries
+
+To improve the performance of transactions, CockroachDB include a set of statements that let you retry transaction that fail due to contention errors. Retrying transactions has the benefit of increasing their priority each time they're retried, increasing their likelihood to succeed.
+
+Implementing client-side retries requires three statements:
+
+- [`SAVEPOINT cockroach_restart`](savepoint.html) defines your intent to retry the transaction if there are contention errors. It must be executed after `BEGIN` but before the first statement that manipulates a database.
+- [`ROLLBACK TO cockroach_restart`](rollback-transaction.html#retry-a-transaction) is used when your application detects `40001` / `retry transaction` errors. It provides you a chance to "retry" the transaction by rolling the database's state back to the beginning of the transaction and incrementing the transaction's priority. 
+  
+  After issuing `ROLLBACK TO cockroach_restart`, each of the transaction's statements must then be executed again.
+
+- [`RELEASE SAVEPOINT cockroach_restart`](release-savepoint.html) commits the transaction.
```

Comment:
a think a note here explaining why this RELEASE is necessary would be helpful: it gives the client the chance to trap a "retryable" error that can happen when attempting to commit. And, in fact, when attempting to commit is one of the more likely places when the client might find out that it needs to retry, because of implementation details: sometimes the transaction finds out late that it has been "pushed" by another one. And sometimes we even let transactions that we know will have to be eventually restarted run for a while, so they lay out more intents and thus have a higher chance of succeeding the next time around.

And the thing is `COMMIT` does not allow a client to receive this errors and act on them. For one, it's often not under the control of the client, as it's issued by a driver. And second, there's no recourse from COMMIT, at least not with the postgres protocol.